### PR TITLE
Add group-by for lists and agentsets

### PR DESCRIPTION
**Pull request by qiemem**
One primitive, can take either a list or agentset.

### CODE REVIEW

File: documentation.conf [hunk 1]
```diff
@@ -48,6 +48,32 @@ primitives = [
     description: "Counts the occurrences of each element of the given list and reports the counts in a table."
   },
   {
+    name: group-by,
+    type: reporter,
+    returns: table,
+    arguments: [ { type: agentset-or-list }],
```
Comment (mrerrormessage) [inline documentation.conf]: Only one argument is listed here, this also takes an anonymous reporter.

File: src/TableExtension.java [hunk 2]
```diff
@@ -451,6 +454,47 @@ public Object report(Argument args[], Context context) throws ExtensionException
...
+        AgentIterator agentIter = agents.shufflerator(context.getRNG());
+        while (agentIter.hasNext()) {
+          org.nlogo.agent.Agent agent = agentIter.next();
+          childContext.agent = agent;
+          Object group = classifier.report(childContext, new Object[0]);
+          ensureKeyValidity(group);
+          res.computeIfAbsent(group, k -> new ArrayList<>()).add(agent);
+        }
+        res.forEach((k, v) ->
+                result.put(k, AgentSet.fromArray(agents.kind(), v.toArray(new org.nlogo.agent.Agent[v.size()])))
```
Comment (qiemem) [inline src/TableExtension.java]: Doh, only because I forgot about `AgentSetBuilder` as I was writing this.

**Comment by qiemem** [comment]
> Blocks can make use of the blockAgentClassString compile-time checking to make sure that the code within the block can be run by the executing agent.

I figured this was more acceptable since this is in an extension.

> It prevents the somewhat confusing situation where code is evaluated in both a different lexical location than where it appears and a different agent context than suggested by the lexical context of its creation.

I guess I don't see how this is more confusing than just switching lexical location. The behavior of e.g. `myself` is the one thing I can see being weird (since `myself` isn't closed over), but I would consider this significantly less confusing than, say, the behavior of `stop` and `report` in closures.

I don't understand your point error reporting. It seems like this same problem exists with all lambdas that are called from a different location than they are created.

> We are evaluating an anonymous reporter where the argument is the same as self (and so `[ x -> x = self ]` is always true). This means that `[ pycor ] and [ p -> [pycor] of p]` represent the same anonymous reporter.

When run on an agent, no argument is passed to reporter, so this is not true.

> Because this alters self, it also alters myself and other-end. `[ p -> myself = p ]` now means group-by the predicate of whether each agent in the set is the same as the agent which calls group-by. I consider this behavior surprising and complex.

As before, `[ p -> myself = p ]` doesn't work as the agent is not passed in as an argument. As for setting `myself`, etc, this seems quite desirable, and is exactly how the primitive would work with a block.

> NetLogo has two evaluate-each-element sort primitives

I've always considered this to be a mistake, tbh. Changing whether we're sorting lists or agents shouldn't change whether we're sorting with a key or comparator.

> This also means that this primitive will return dramatically different results when run on a list of agents than it will when run on an agentset when the anonymous reporter references any variable of the calling agent, `self`, `myself`, or `other-end`.

That's true. I hadn't considered that. However, I don't consider this to be much of a problem because attempting to call a list grouper as an agentset grouper will result in an error (since there's no argument passed in).


All that said, while I prefer the given solution, I would be okay with adding a separate `table:group-agents-by` (I don't like `table:group-on` exactly because of the confusion commonly produced by `sort-by` vs `sort-on`; tbh I still have to look up the sorting procedures every time I want to use them).

**Comment by mrerrormessage** [comment]
Looks good. Thanks for making this!

**Pull request merged**